### PR TITLE
csi: enable csi-addons-side when crds are deployed

### PR DESCRIPTION
**Pull request by subhamkrai**
**Description of your changes:**
when deploying csi, we'll check if csi-addons crds are deployed we'll enable csi-addons sidecar by default.

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by subhamkrai** [comment]
I don't want to rush this pr or may be even close this depending on what we decide. I just created this what I had in mind and before I forgot since I'll off sometime next week.

JFYI, I did quick test with csi-addons crd present but sidecar was false in configmap but after the reconcile, I see sidecar was started and network fence was created.

**Comment by subhamkrai** [comment]
@Madhu-1 @travisn do you think these are valid changes to accept?

File: pkg/operator/ceph/csi/csi.go [hunk 1]
```diff
@@ -179,6 +179,10 @@ func (r *ReconcileCSI) setParams(ver *version.Info) error {
 	if strings.EqualFold(k8sutil.GetValue(r.opConfig.Parameters, "CSI_ENABLE_CSIADDONS", "false"), "true") {
 		CSIParam.EnableCSIAddonsSideCar = true
 	}
+	_, err = r.context.ApiExtensionsClient.ApiextensionsV1().CustomResourceDefinitions().Get(r.opManagerContext, "networkfences.csiaddons.openshift.io", metav1.GetOptions{})
+	if err == nil {
+		CSIParam.EnableCSIAddonsSideCar = true
+	}
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/csi.go:185]:
```suggestion
CSIParam.EnableCSIAddonsSideCar = false
	_, err = r.context.ApiExtensionsClient.ApiextensionsV1().CustomResourceDefinitions().Get(r.opManagerContext, "networkfences.csiaddons.openshift.io", metav1.GetOptions{})
	if err == nil {
		CSIParam.EnableCSIAddonsSideCar = true
	}
if strings.EqualFold(k8sutil.GetValue(r.opConfig.Parameters, "CSI_ENABLE_CSIADDONS", ""), "false") {
		CSIParam.EnableCSIAddonsSideCar = false
	}
```

i think above might be the good as we still allow users to disable csi_addons if they want

Comment (travisn) [reply pkg/operator/ceph/csi/csi.go:185]: Agreed, this looks good as long as the user has the ability to disable the sidecar if desired.

**Review by travisn — CHANGES_REQUESTED** [review]
note the csi unit tests are failing

File: pkg/operator/ceph/csi/csi.go [hunk 2]
```diff
@@ -176,9 +176,13 @@ func (r *ReconcileCSI) setParams(ver *version.Info) error {
...
 
 	CSIParam.EnableCSIAddonsSideCar = false
-	if strings.EqualFold(k8sutil.GetValue(r.opConfig.Parameters, "CSI_ENABLE_CSIADDONS", "false"), "true") {
+	_, err = r.context.ApiExtensionsClient.ApiextensionsV1().CustomResourceDefinitions().Get(r.opManagerContext, "csiaddonsnode.csiaddons.openshift.io", metav1.GetOptions{})
+	if err == nil {
 		CSIParam.EnableCSIAddonsSideCar = true
 	}
+	if strings.EqualFold(k8sutil.GetValue(r.opConfig.Parameters, "CSI_ENABLE_CSIADDONS", ""), "false") {
+		CSIParam.EnableCSIAddonsSideCar = false
+	}
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/csi.go:185]: I think we are missing one case here

* Enable the sidecar first and install the csiaddons CRD after sometime.

With current change we will never enable the sidecar

Comment (subhamkrai) [reply pkg/operator/ceph/csi/csi.go:185]: > Enable the sidecar first and install the csiaddons CRD after sometime.

any suggestion on how to check the time after which we should install csiaddons?

Comment (subhamkrai) [reply pkg/operator/ceph/csi/csi.go]: done

File: pkg/operator/ceph/csi/controller.go [hunk 3]
```diff
@@ -89,6 +90,11 @@ func add(ctx context.Context, mgr manager.Manager, r reconcile.Reconciler, opCon
 	}
 	logger.Infof("%s successfully started", controllerName)
 
+	err = addonsv1alpha1.AddToScheme(mgr.GetScheme())
```
Comment (parth-gr) [inline pkg/operator/ceph/csi/controller.go:93]: can you add a comment what exactly we are doing here

**Review by travisn — APPROVED** [review]

File: pkg/operator/ceph/csi/controller.go (hunk 3 above)
Comment (subhamkrai) [reply pkg/operator/ceph/csi/controller.go:94]: added

File: pkg/operator/ceph/csi/csi.go [hunk 4]
```diff
@@ -175,8 +175,14 @@ func (r *ReconcileCSI) setParams(ver *version.Info) error {
 		CSIParam.EnableNFSSnapshotter = false
 	}
 
-	CSIParam.EnableCSIAddonsSideCar = false
-	if strings.EqualFold(k8sutil.GetValue(r.opConfig.Parameters, "CSI_ENABLE_CSIADDONS", "false"), "true") {
```
Comment (Madhu-1) [inline pkg/operator/ceph/csi/csi.go:179]: `CSIParam.EnableCSIAddonsSideCar = false` let's keep this one as it is the default value for the Reconcile

**Review by parth-gr — APPROVED** [review]

File: pkg/operator/ceph/csi/csi.go (hunk 4 above)
Comment (subhamkrai) [reply pkg/operator/ceph/csi/csi.go:179]: done

**Review by Madhu-1 — APPROVED** [review]

**Pull request merged**